src/input.rs

```diff
@@ -729,6 +738,18 @@ impl<'a, A: ActionContext + 'a> Processor<'a, A> {
                 },
                 ElementState::Released => self.on_mouse_release(button, modifiers, point),
             }
+
+            let button_changed = self.ctx.mouse().last_button != button;
+            let elapsed = self.ctx.mouse().last_click_timestamp.elapsed();
+
+            match self.ctx.mouse().click_state {
+                ClickState::DoubleClick
+                    if !button_changed && elapsed < self.mouse_config.double_click.threshold =>
+                {
+                    self.on_mouse_double_click(button, state, point);
+                }
+                _ => {}
+            }
```

Comment:
The problem I see is that you want a couple different steps to be taken on a mouse action:

1. set the click state according to the current click state and the current mouse button state (e.g. `ClickState::Click` + `ElementState::Pressed` == `ClickState::DoubleClick`)
2. then, you want to perform any action related to the new `ClickState`, plus potentially the current `ElementState` (such as executing any of the possible selection types)

`on_mouse_press` is (mostly) now performing the first step, and all the selections from the second step, since all of those are triggered on `ElementState::Pressed`.

This new selection type doesn't fit that model, since it should act on `Released`, not pressed.

I don't yet see a very clean way to fit that into the current code, without pulling apart the `on_mouse_press` and basically split it into a function that _only_ does (1), and then separate functions that trigger based on the current `ClickState` and `ElementState`, after that, to resolve (2).

I'm not quite sure what you'd prefer. I worked off of this comment you gave before:

> I think it might make sense to move the `on_mouse_double_click` and `on_mouse_triple_click` into the `mouse_input` function and then just passing the pressed/released state as arguments. That should clean stuff up and resolve the issue that you mentioned with expecting the double click method to be called on release too.

But I didn't want to start refactoring all of it just yet.

There are several ways to represent the end-goal, I have no preference for either of them, but if you have an idea of what you'd like to see, let me know and I'll implement that.

Reply:
I'd have to play around with the code myself to see which exact limitations apply, but that would likely take quite a while. So going off the comment I gave earlier, I think that would still make sense.

Using the current functions as single press functions seems fine to me. Then we can just extract the `on_mouse_double_click` and `on_mouse_triple_click` functions into the the `mouse_input` function. So we split single/double/triple click first and then by press/release, instead of the other way around.

However I'm open for alternative solutions. Basically the only thing that doesn't seem like a good solution is the way it's in right now. I don't think there's any way around refactoring. We need our code to stay structured to be able to find stuff and the way it is right now, I don't think that's the case.

If you have a better idea than my proposal, I'm open to that too. I just don't think there's any way around refactoring things at least a little bit. Since as you've mentioned yourself the functionality just doesn't fit into the current structure very well. Tacking it on as a separate code path just doesn't feel like the right approach.